src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs

```diff
@@ -135,93 +135,174 @@ private void LoadTestsFromAnExtension(string extensionAssembly, IEnumerable<stri
-                    EqtTrace.Error("DiscovererEnumerator.LoadTestsFromAnExtension: {0} ", e);
+            if (this.discoveryResultCache.TotalDiscoveredTests == 0)
+            {
+                DiscovererEnumerator.LogWarningOnNoTestsDiscovered(sources, testCaseFilter, logger);
+            }
 
-                    continue;
-                }
+            this.CollectTelemetryAtEnd(totalTimeTakenByAdapters, totalAdaptersUsed);
+        }
+
+        private void CollectTelemetryAtEnd(double totalTimeTakenByAdapters, double totalAdaptersUsed)
+        {
+            // Collecting Total Time Taken by Adapters
+            this.requestData.MetricsCollection.Add(TelemetryDataConstants.TimeTakenInSecByAllAdapters,
+                totalTimeTakenByAdapters);
+
+            // Collecting Total Adapters Used to Discover tests
+            this.requestData.MetricsCollection.Add(TelemetryDataConstants.NumberOfAdapterUsedToDiscoverTests,
+                totalAdaptersUsed);
+        }
+
+        private void DiscoverTestsFromSingleDiscoverer(
+            LazyExtension<ITestDiscoverer, ITestDiscovererCapabilities> discoverer,
+            Dictionary<LazyExtension<ITestDiscoverer, ITestDiscovererCapabilities>,IEnumerable<string>> discovererToSourcesMap,
+            DiscoveryContext context,
+            TestCaseDiscoverySink discoverySink,
+            IMessageLogger logger,
+            ref bool discoverersFromDeprecatedLocations,
+            ref double totalAdaptersUsed,
+            ref double totalTimeTakenByAdapters)
+        {
+            if (TryToLoadDiscoverer(discoverer, logger, out var discovererType) == false)
+            {
+                // Fail to instantiate the discoverer type.
+                return;
+            }
+
+            // on instantiated successfully, get tests
+            try
+            {
+                EqtTrace.Verbose(
+                    "DiscovererEnumerator.DiscoverTestsFromSingleDiscoverer: Loading tests for {0}",
+                        discoverer.Value.GetType().FullName);
+
+                var currentTotalTests = this.discoveryResultCache.TotalDiscoveredTests;
+                var newTimeStart = DateTime.UtcNow;
 
-                // if instantiated successfully, get tests
-                try
+                this.testPlatformEventSource.AdapterDiscoveryStart(discoverer.Metadata.DefaultExecutorUri.AbsoluteUri);
+                discoverer.Value.DiscoverTests(discovererToSourcesMap[discoverer], context, logger, discoverySink);
+
+                var totalAdapterRunTime = DateTime.UtcNow - newTimeStart;
+
+                this.testPlatformEventSource.AdapterDiscoveryStop(this.discoveryResultCache.TotalDiscoveredTests -
+                                                                  currentTotalTests);
+
+                // Record Total Tests Discovered By each Adapter.
+                this.RecordTotalTestsDiscoveredByCurrentAdapter(
```

Comment:
Better naming? May be `RecordTotalTestsDiscoveredByDiscoverer` as we are already passing discoverer

Reply:
After removing refs. this method no more required.